src/components/swap/SwapRoute.tsx

```diff
@@ -1,47 +1,88 @@
-}
+import { FeeAmount, Trade as V3Trade } from '@uniswap/v3-sdk'
+import { AutoColumn } from 'components/Column'
+import { LoadingRows } from 'components/Loader/styled'
+import RoutingDiagram, { RoutingDiagramEntry } from 'components/RoutingDiagram/RoutingDiagram'
+import { RowBetween } from 'components/Row'
+import { Version } from 'hooks/useToggledVersion'
+import { memo } from 'react'
+import styled from 'styled-components/macro'
+import { TYPE } from 'theme'
+import { getTradeVersion } from 'utils/getTradeVersion'
+import { RouterLabel } from './RouterLabel'
+
+const Separator = styled.div`
+  border-top: 1px solid ${({ theme }) => theme.bg2};
+  width: 100%;
+  height: 1px;
+`
+
+const V2_DEFAULT_FEE_TIER = 3000
 
 export default memo(function SwapRoute({
   trade,
+  loading,
 }: {
   trade: V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType>
+  loading: boolean
 }) {
-  const tokenPath = trade instanceof V2Trade ? trade.route.path : trade.route.tokenPath
-  const theme = useContext(ThemeContext)
   return (
-    <Flex flexWrap="wrap" width="100%" justifyContent="flex-start" alignItems="center">
-      {tokenPath.map((token, i, path) => {
-        const isLastItem: boolean = i === path.length - 1
-        const currency = unwrappedToken(token)
-        return (
-          <Fragment key={i}>
-            <Flex alignItems="end">
-              <TYPE.black color={theme.text1} ml="0.145rem" mr="0.145rem">
-                {currency.symbol}
-              </TYPE.black>
-            </Flex>
-            {isLastItem ? null : trade instanceof V2Trade ? (
-              <ChevronRight size={14} color={theme.text2} />
-            ) : (
-              <LabeledArrow fee={trade.route.pools[i].fee} />
-            )}
-          </Fragment>
-        )
-      })}
-    </Flex>
+    <AutoColumn gap="12px">
+      <RowBetween>
+        <RouterLabel />
+        {!loading && (
+          <TYPE.black fontSize={14}>
+            {getTradeVersion(trade) === Version.v2 ? <Trans>via V2</Trans> : <Trans>via V3</Trans>}
+          </TYPE.black>
+        )}
+      </RowBetween>
+      <Separator />
+      {loading ? (
+        <LoadingRows>
+          <div style={{ width: '400px', height: '30px' }} />
+        </LoadingRows>
+      ) : (
+        <RoutingDiagram
+          currencyIn={trade.inputAmount.currency}
+          currencyOut={trade.outputAmount.currency}
+          routes={getTokenPath(trade)}
+        />
+      )}
+    </AutoColumn>
   )
 })
+
+function getTokenPath(
+  trade: V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType>
+): RoutingDiagramEntry[] {
+  // convert V2 path to a list of routes
+  if (trade instanceof V2Trade) {
+    const { path: tokenPath } = (trade as V2Trade<Currency, Currency, TradeType>).route
+    const path = []
+    for (let i = 1; i < tokenPath.length; i++) {
+      path.push([tokenPath[i - 1], tokenPath[i], V2_DEFAULT_FEE_TIER] as RoutingDiagramEntry['path'][0])
```

Comment:
Not familiar w/ the `Type['key'][n]` syntax here. Is this for sure the best way to type the things it's being applied to?

Reply:
hm that just makes it so that i don't have to extract the inner types to an external variable. That type doesn't really have meaning outside of the super type.